hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bucket/SparkHiveBucketIndex.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.index.bucket;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hudi.client.WriteStatus;
+import org.apache.hudi.common.engine.HoodieEngineContext;
+import org.apache.hudi.common.model.HoodieBaseFile;
+import org.apache.hudi.common.model.HoodieKey;
+import org.apache.hudi.common.model.HoodieRecord;
+import org.apache.hudi.common.model.HoodieRecordLocation;
+import org.apache.hudi.common.model.HoodieRecordPayload;
+import org.apache.hudi.common.table.HoodieTableConfig;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.collection.Triple;
+import org.apache.hudi.config.HoodieIndexConfig;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.exception.HoodieIndexException;
+import org.apache.hudi.index.HoodieIndexUtils;
+import org.apache.hudi.index.SparkHoodieIndex;
+import org.apache.hudi.table.HoodieTable;
+import org.apache.hudi.utils.BucketUtils;
+import org.apache.spark.api.java.JavaRDD;
+
+public class SparkHiveBucketIndex<T extends HoodieRecordPayload> extends SparkHoodieIndex<T> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SparkHiveBucketIndex.class);
+
+  private final int numBuckets;
+
+  public SparkHiveBucketIndex(HoodieWriteConfig config) {
+    super(config);
+    String tableNumBucket = config.getProps()
+        .getProperty(HoodieTableConfig.HOODIE_TABLE_NUM_BUCKETS.key());
+    String indexNumBucket = config.getProps()
+        .getProperty(HoodieIndexConfig.BUCKET_INDEX_BUCKET_NUM.key());
+    if (tableNumBucket == null && indexNumBucket == null) {
+      throw new IllegalArgumentException(
+          "Please set hoodie.index.hive.bucket.num or hoodie.table.numbuckets(deprecated) to a positive integer.");
```

Comment:
hoodie.table.numbuckets is not exist? and please change `hoodie.index.hive.bucket.num` hardcoded to config.

Reply:
`hoodie.table.numbuckets` is in the table config. Bucket info stored as a part of table meta data is beneficial to optimizing the query plan. The value of this config has to be the same with `HoodieIndexConfig.HIVE_BUCKET_INDEX_BUCKET_NUM`

Reply:
Besides, bucket num of a table is always a fixed value and rarely change

Reply:
I prefer to keep the `hoodie.table.numbuckets` instead of `hoodie.index.hive.bucket.num`

Reply:
At last, `hoodie.index.hive.bucket.num` is used to get the table bucket num instead of 
 being stored in hoodie.properties. It is hard to access. Hive metastore server can be used to store the table bucket info.